rfcs/20201103-binary-bcast-source0/README.md

```diff
@@ -0,0 +1,74 @@
+# Proposal for enable broadcast source 0 in binary
+
+## Problem and motivation
+Currently in OneDNN library binary primitive supports implicit broadcast 
+semantics for source 1. It means that if some dimension has value of one, 
+this value will be used to compute an operation with each point of 
+source 0 for this dimension. There is also a need for enabling this functionality
+for source 1 to handle binary operation on sources like `Nx1` and `1xM`. This 
+will improve oneDNN integration in TensorFlow library. At this point, was 
+reported, that performance (jit kernel) is not the main concern.
+
+![](bcast_example.png "Binary ADD operation 1x4:4x1")
+
+## Overview
+
+### Current implementations
+*  Reference implementation allows broadcast with following syntax:
+   `NxCxDxHxW:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+*  JIT implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{N,1}xCx1x1x1 -> NxCxDxHxW`
+### New implementations
+*  Reference implementation allows broadcast with following syntax:
+   `N/1xC/1xD/1xH/1xW/1:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+### Possible future implementations (if the need is reported)
+*  JIT implementation allows broadcast with following syntax:
+   `{NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1}:
+    {NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `{N,1}xCx{D,1}x{H,1}x{W,1}:{N,1}xCx{D,1}x{H,1}x{W,1} -> NxCxDxHxW`
+
+Similarly for other formats.
+
+## Proposal
+The proposal is to extend binary primitive by removing a restriction 
+on src0 broadcasting. The reference implementation will be the only 
+implementation supporting this broadcasting until we have an additional 
+request. The dimensions of the destination should be defined as: 
+`dst_dims[d] = max(src0_dims[d], src1_dims[d])` and 
+`src0_dims[d] == src1_dims[d] || one_of(1, src0_dims[d], src1_dims[d])`. Thanks 
+to this, we will get backward compatibility. If user doesnt use broadcast on 
+source0, source0 dimensions are the same as dst. The post-ops operations also 
+remain unchanged. If it will be reported, can also add support for broadcast 0
+in jit kernels.
+
+### Post-ops handle
+
+We currently have two options on how to handle post-ops with this change:
+
+1. Add a limitation to binary post operation so dst_dims should be equal 
+   to src0_dims. (probably should be chosen as the need is not requested 
+   by users and to avoid complications in post-ops)
```

Comment:
```suggestion
1. Restrict binary post-op src1 md (input) to have dimension values greater then dst_md of original operation. This means that broadcast operation for src0 in case of binary post-op is prohibited, but broadcast for src1 will still be supported as is.
```